File: tests/src/main/java/com/netflix/msl/util/MslTestUtils.java
/**
 * Copyright (c) 2012-2020 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.msl.util;

import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.Date;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

import javax.crypto.Mac;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

import com.netflix.msl.MslCryptoException;
import com.netflix.msl.MslEncodingException;
import com.netflix.msl.MslException;
import com.netflix.msl.MslInternalException;
import com.netflix.msl.crypto.ICryptoContext;
import com.netflix.msl.crypto.JcaAlgorithm;
import com.netflix.msl.crypto.NullCryptoContext;
import com.netflix.msl.entityauth.EntityAuthenticationData;
import com.netflix.msl.entityauth.MockPresharedAuthenticationFactory;
import com.netflix.msl.io.MslEncodable;
import com.netflix.msl.io.MslEncoderException;
import com.netflix.msl.io.MslEncoderFactory;
import com.netflix.msl.io.MslObject;
import com.netflix.msl.tokens.MasterToken;
import com.netflix.msl.tokens.MslUser;
import com.netflix.msl.tokens.ServiceToken;
import com.netflix.msl.tokens.UserIdToken;

/**
 * Helper functions common to many unit tests and mocks.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
public class MslTestUtils {
    /** Base service token name. */
    private static final String SERVICE_TOKEN_NAME = "serviceTokenName";
    /**
     * Maximum number of service tokens to randomly generate. This needs to be
     * large enough to statistically create the applicable set of service
     * tokens for the tests.
     */
    private static final int NUM_SERVICE_TOKENS = 12;
    
    /** Wrapping key derivation algorithm salt. */
    private static final byte[] SALT = {
        (byte)0x02, (byte)0x76, (byte)0x17, (byte)0x98, (byte)0x4f, (byte)0x62, (byte)0x27, (byte)0x53, 
        (byte)0x9a, (byte)0x63, (byte)0x0b, (byte)0x89, (byte)0x7c, (byte)0x01, (byte)0x7d, (byte)0x69 };
    /** Wrapping key derivation algorithm info. */
    private static final byte[] INFO = {
        (byte)0x80, (byte)0x9f, (byte)0x82, (byte)0xa7, (byte)0xad, (byte)0xdf, (byte)0x54, (byte)0x8d,
        (byte)0x3e, (byte)0xa9, (byte)0xdd, (byte)0x06, (byte)0x7f, (byte)0xf9, (byte)0xbb, (byte)0x91, };
    /** Wrapping key length in bytes. */
    private static final int WRAPPING_KEY_LENGTH = 128 / Byte.SIZE;

    
    /**
     * Parse a new {@link MslObject} from the {@link MslEncodable}.
     * 
     * @param encoder the {@link MslEncoderFactory}.
     * @param encode a {@link MslEncodable}.
     * @return the {@link MslObject}
     * @throws MslEncoderException if there is an error encoding and converting
     *         the  object cannot be encoded and converted
     */
    public static MslObject toMslObject(final MslEncoderFactory encoder, final MslEncodable encode) throws MslEncoderException {
        final byte[] encoding = encode.toMslEncoding(encoder, encoder.getPreferredFormat(null));
        return encoder.parseObject(encoding);
    }

    /**
     * Returns a master token with the identity of the MSL context entity
     * authentication data that is not renewable or expired.
     * 
     * @param ctx MSL context.
     * @param sequenceNumber master token sequence number to use.
     * @param serialNumber master token serial number to use.
     * @return a new master token.
     * @throws MslEncodingException if there is an error encoding the JSON
     *         data.
     * @throws MslCryptoException if there is an error encrypting or signing
     *         the token data.
     */
    public static MasterToken getMasterToken(final MslContext ctx, final long sequenceNumber, final long serialNumber) throws MslEncodingException, MslCryptoException {
        final Date renewalWindow = new Date(System.currentTimeMillis() + 10000);
        final Date expiration = new Date(System.currentTimeMillis() + 20000);
        final EntityAuthenticationData entityAuthData = ctx.getEntityAuthenticationData(null);
        final String identity = entityAuthData.getIdentity();
        final SecretKey encryptionKey = MockPresharedAuthenticationFactory.KPE;
        final SecretKey hmacKey = MockPresharedAuthenticationFactory.KPH;
        return new MasterToken(ctx, renewalWindow, expiration, sequenceNumber, serialNumber, null, identity, encryptionKey, hmacKey);
    }
    
    /**
     * Returns an untrusted master token with the identity of the MSL context
     * entity authentication data that is not renewable or expired.
     * 
     * @param ctx MSL context.
     * @return a new untrusted master token.
     * @throws MslEncodingException if there is an error encoding the data.
     * @throws MslCryptoException if there is an error encrypting or signing
     *         the token data.
     * @throws MslException if the master token is constructed incorrectly.
     * @throws MslEncoderException if there is an error editing the data.
     */
    public static MasterToken getUntrustedMasterToken(final MslContext ctx) throws MslEncodingException, MslCryptoException, MslEncoderException, MslException {
        final Date renewalWindow = new Date(System.currentTimeMillis() + 10000);
        final Date expiration = new Date(System.currentTimeMillis() + 20000);
        final EntityAuthenticationData entityAuthData = ctx.getEntityAuthenticationData(null);
        final String identity = entityAuthData.getIdentity();
        final SecretKey encryptionKey = MockPresharedAuthenticationFactory.KPE;
        final SecretKey hmacKey = MockPresharedAuthenticationFactory.KPH;
        final MasterToken masterToken = new MasterToken(ctx, renewalWindow, expiration, 1L, 1L, null, identity, encryptionKey, hmacKey);
        final MslEncoderFactory encoder = ctx.getMslEncoderFactory();
        final MslObject mo = toMslObject(encoder, masterToken);
        final byte[] signature = mo.getBytes("signature");
        ++signature[1];
        mo.put("signature", signature);
        return new MasterToken(ctx, mo);
    }
    
    /**
     * Returns a user ID token with the identity of the provided user that is
     * not renewable or expired.
     * 
     * @param ctx MSL context.
     * @param masterToken master token to bind against.
     * @param serialNumber user ID token serial number to use.
     * @param user MSL user to use.
     * @return a new user ID token.
     * @throws MslEncodingException if there is an error encoding the JSON
     *         data.
     * @throws MslCryptoException if there is an error encrypting or signing
     *         the token data.
     */
    public static UserIdToken getUserIdToken(final MslContext ctx, final MasterToken masterToken, final long serialNumber, final MslUser user) throws MslEncodingException, MslCryptoException {
        final Date renewalWindow = new Date(System.currentTimeMillis() + 10000);
        final Date expiration = new Date(System.currentTimeMillis() + 20000);
        return new UserIdToken(ctx, renewalWindow, expiration, masterToken, serialNumber, null, user);
    }
    
    /**
     * Returns an untrusted user ID token with the identity of the provided
     * user that is not renewable or expired.
     * 
     * @param ctx MSL context.
     * @param masterToken master token to bind against.
     * @param serialNumber user ID token serial number to use.
     * @param user MSL user to use.
     * @return a new untrusted user ID token.
     * @throws MslEncodingException if there is an error encoding the data.
     * @throws MslCryptoException if there is an error encrypting or signing
     *         the token data.
     * @throws MslEncoderException if there is an error editing the data.
     * @throws MslException if the user ID token serial number is out of range.
     */
    public static UserIdToken getUntrustedUserIdToken(final MslContext ctx, final MasterToken masterToken, final long serialNumber, final MslUser user) throws MslEncodingException, MslCryptoException, MslEncoderException, MslException {
        final Date renewalWindow = new Date(System.currentTimeMillis() + 10000);
        final Date expiration = new Date(System.currentTimeMillis() + 20000);
        final UserIdToken userIdToken = new UserIdToken(ctx, renewalWindow, expiration, masterToken, serialNumber, null, user);
        final MslEncoderFactory encoder = ctx.getMslEncoderFactory();
        final MslObject mo = toMslObject(encoder, userIdToken);
        final byte[] signature = mo.getBytes("signature");
        ++signature[1];
        mo.put("signature", signature);
        return new UserIdToken(ctx, mo, masterToken);
    }
    
    /**
     * @param ctx MSL context.
     * @param masterToken master token to bind against. May be null.
     * @param userIdToken user ID token to bind against. May be null.
     * @return a set of new service tokens with random token bindings.
     * @throws MslEncodingException if there is an error encoding the JSON
     *         data.
     * @throws MslCryptoException if there is an error encrypting or signing
     *         the token data.
     * @throws MslException if there is an error compressing the data.
     */
    public static Set<ServiceToken> getServiceTokens(final MslContext ctx, final MasterToken masterToken, final UserIdToken userIdToken) throws MslEncodingException, MslCryptoException, MslException {
        final Random random = new Random();
        final ICryptoContext cryptoContext = new NullCryptoContext();
        final Set<ServiceToken> serviceTokens = new HashSet<ServiceToken>();
        final int numTokens = Math.max(NUM_SERVICE_TOKENS, 3);
        for (int i = 0; i < numTokens; ++i) {
            final String name = SERVICE_TOKEN_NAME + random.nextInt();
            final byte[] data = new byte[32];
            random.nextBytes(data);

            // Make sure one of each type of token is included.
            // Otherwise pick a random type.
            final int type = (i < 3) ? i : random.nextInt(3);
            switch (type) {
                case 0:
                    serviceTokens.add(new ServiceToken(ctx, name, data, null, null, false, null, cryptoContext));
                    break;
                case 1:
                    serviceTokens.add(new ServiceToken(ctx, name, data, masterToken, null, false, null, cryptoContext));
                    break;
                case 2:
                    serviceTokens.add(new ServiceToken(ctx, name, data, masterToken, userIdToken, false, null, cryptoContext));
                    break;
            }
        }
        return serviceTokens;
    }
    
    /**
     * @param ctx MSL context.
     * @param masterToken the master token to bind against.
     * @return a random set of master token bound service tokens.
     * @throws MslEncodingException if there is an error constructing the
     *         service token.
     * @throws MslCryptoException if there is an error constructing the service
     *         token.
     * @throws MslException if there is an error compressing the data.
     */
    public static Set<ServiceToken> getMasterBoundServiceTokens(final MslContext ctx, final MasterToken masterToken) throws MslEncodingException, MslCryptoException, MslException {
        final Random random = new Random();
        final ICryptoContext cryptoContext = new NullCryptoContext();
        final Set<ServiceToken> tokens = new HashSet<ServiceToken>();
        for (int count = random.nextInt(NUM_SERVICE_TOKENS); count >= 0; --count) {
            final String name = SERVICE_TOKEN_NAME + random.nextInt();
            final byte[] data = new byte[8];
            random.nextBytes(data);
            final ServiceToken token = new ServiceToken(ctx, name, data, masterToken, null, false, null, cryptoContext);
            tokens.add(token);
        }
        return tokens;
    }
    
    /**
     * @param ctx MSL context.
     * @param masterToken the master token to bind against.
     * @param userIdToken the user ID token to bind against.
     * @return a random set of user ID token bound service tokens.
     * @throws MslEncodingException if there is an error constructing the
     *         service token.
     * @throws MslCryptoException if there is an error constructing the service
     *         token.
     * @throws MslException if there is an error compressing the data.
     */
    public static Set<ServiceToken> getUserBoundServiceTokens(final MslContext ctx, final MasterToken masterToken, final UserIdToken userIdToken) throws MslEncodingException, MslCryptoException, MslException {
        final Random random = new Random();
        final ICryptoContext cryptoContext = new NullCryptoContext();
        final Set<ServiceToken> tokens = new HashSet<ServiceToken>();
        for (int count = random.nextInt(NUM_SERVICE_TOKENS); count >= 0; --count) {
            final String name = SERVICE_TOKEN_NAME + random.nextInt();
            final byte[] data = new byte[8];
            random.nextBytes(data);
            final ServiceToken token = new ServiceToken(ctx, name, data, masterToken, userIdToken, false, null, cryptoContext);
            tokens.add(token);
        }
        return tokens;
    }
    
    /**
     * Derives the pre-shared or model group keys AES-128 Key Wrap key from the
     * provided AES-128 encryption key and HMAC-SHA256 key.
     * 
     * @param encryptionKey the encryption key.
     * @param hmacKey the HMAC key.
     * @return the wrapping key.
     */
    public static byte[] deriveWrappingKey(final byte[] encryptionKey, final byte[] hmacKey) {
        try {
            // Concatenate the keys.
            final byte[] bits = Arrays.copyOf(encryptionKey, encryptionKey.length + hmacKey.length);
            System.arraycopy(hmacKey, 0, bits, encryptionKey.length, hmacKey.length);
    
            // HMAC-SHA256 the keys with the salt as the HMAC key.
            final SecretKey saltKey = new SecretKeySpec(SALT, JcaAlgorithm.HMAC_SHA256);
            final Mac intermediateMac = Mac.getInstance(saltKey.getAlgorithm());
            intermediateMac.init(saltKey);
            final byte[] intermediateBits = intermediateMac.doFinal(bits);
    
            // HMAC-SHA256 the info with the intermediate key as the HMAC key.
            final SecretKey intermediateKey = new SecretKeySpec(intermediateBits, JcaAlgorithm.HMAC_SHA256);
            final Mac finalMac = Mac.getInstance(intermediateKey.getAlgorithm());
            finalMac.init(intermediateKey);
            final byte[] finalBits = finalMac.doFinal(INFO);

            // Grab the first 128 bits.
            return Arrays.copyOf(finalBits, WRAPPING_KEY_LENGTH);
        } catch (final NoSuchAlgorithmException e) {
            throw new MslInternalException(JcaAlgorithm.HMAC_SHA256 + " algorithm not found.", e);
        } catch (final InvalidKeyException e) {
            throw new MslInternalException("Invalid " + JcaAlgorithm.HMAC_SHA256 + " key.", e);
        }
    }
}

File: tests/src/main/cpp/util/MslTestUtils.cpp
/**
 * Copyright (c) 2016-2020 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "MslTestUtils.h"

#include <crypto/IRandom.h>
#include <crypto/NullCryptoContext.h>
#include <Date.h>
#include <crypto/OpenSslLib.h>
#include <entityauth/EntityAuthenticationData.h>
#include <io/MslEncodable.h>
#include <io/MslEncoderFactory.h>
#include <MslCryptoException.h>
#include <MslInternalException.h>
#include <tokens/MasterToken.h>
#include <tokens/MslUser.h>
#include <tokens/ServiceToken.h>
#include <tokens/UserIdToken.h>
#include <util/MslContext.h>
#include <algorithm>
#include <memory>
#include <set>
#include <sstream>
#include <string>
#include <openssl/rsa.h>
#include <openssl/x509.h>

#include <gtest/gtest.h>
#include <entityauth/MockPresharedAuthenticationFactory.h>
#include <util/ScopedDisposer.h>

using namespace std;
using namespace testing;
using namespace netflix::msl::crypto;
using namespace netflix::msl::entityauth;
using namespace netflix::msl::io;
using namespace netflix::msl::tokens;
using namespace netflix::msl::util;

namespace netflix {

typedef vector<uint8_t> ByteArray;

namespace msl {
namespace util {

namespace {

template<typename T, size_t N> T* begin(T (&arr)[N]) { return &arr[0];     }
template<typename T, size_t N> T* end(T (&arr)[N])   { return &arr[0] + N; }

/** Wrapping key derivation algorithm salt. */
const unsigned char saltAry[] = {
    0x02, 0x76, 0x17, 0x98, 0x4f, 0x62, 0x27, 0x53,
    0x9a, 0x63, 0x0b, 0x89, 0x7c, 0x01, 0x7d, 0x69
};

/** Wrapping key derivation algorithm info. */
static const unsigned char infoAry[] = {
    0x80, 0x9f, 0x82, 0xa7, 0xad, 0xdf, 0x54, 0x8d,
    0x3e, 0xa9, 0xdd, 0x06, 0x7f, 0xf9, 0xbb, 0x91
};

/** Base service token name. */
const string SERVICE_TOKEN_NAME = "serviceTokenName";
/**
 * Maximum number of service tokens to randomly generate. This needs to be
 * large enough to statistically create the applicable set of service
 * tokens for the tests.
 */
const int32_t NUM_SERVICE_TOKENS = 12;

} // namespace anonymous

namespace MslTestUtils {

shared_ptr<MslObject> toMslObject(shared_ptr<MslEncoderFactory> encoder, shared_ptr<MslEncodable> encode)
{
    shared_ptr<ByteArray> encoding = encode->toMslEncoding(encoder, encoder->getPreferredFormat());
    return encoder->parseObject(encoding);
}

pair<PublicKey,PrivateKey> generateRsaKeys(const string& algo, int length)
{
	static const unsigned long publicExponent = 65537;
	uint32_t retryCount = 0;
	static const uint32_t MAX_RETRIES = 4;
	while (true)
	{
		ScopedDisposer<RSA, void, RSA_free> rsa(RSA_generate_key(length, publicExponent, 0, 0));
		if (!rsa.isEmpty())
		{
			if (RSA_check_key(rsa.get()) == 1)
			{
				// public key in SPKI format
				int keyLen = i2d_RSA_PUBKEY(rsa.get(), NULL);
				shared_ptr<ByteArray> spki = make_shared<ByteArray>(keyLen);
				unsigned char * buf = &(*spki)[0];
				if (!i2d_RSA_PUBKEY(rsa.get(), &buf))
					throw MslInternalException("i2d_RSA_PUBKEY failed");
				PublicKey publicKey(spki, algo);

				// private key in PKCS#8 format
				ScopedDisposer<EVP_PKEY, void, EVP_PKEY_free> pkey(EVP_PKEY_new());
				if (pkey.isEmpty())
					throw MslInternalException("EVP_PKEY_new failed");
				if (!EVP_PKEY_set1_RSA(pkey.get(), rsa.get()))
					throw MslInternalException("EVP_PKEY_set1_RSA failed");
				ScopedDisposer<PKCS8_PRIV_KEY_INFO, void, PKCS8_PRIV_KEY_INFO_free> p8inf(EVP_PKEY2PKCS8(pkey.get()));
				if (p8inf.isEmpty())
					throw MslInternalException("PKCS8_PRIV_KEY_INFO_free failed");
				int outLen = i2d_PKCS8_PRIV_KEY_INFO(p8inf.get(), NULL);
				if (outLen <= 0)
					throw MslInternalException("i2d_PKCS8_PRIV_KEY_INFO failed");
				shared_ptr<ByteArray> pkcs8 = make_shared<ByteArray>(outLen);
				buf = &(*pkcs8)[0];
				if (!i2d_PKCS8_PRIV_KEY_INFO(p8inf.get(), &buf))
					throw MslInternalException("i2d_PKCS8_PRIV_KEY_INFO failed");
				PrivateKey privateKey(pkcs8, algo);

				// done with key generation
				return make_pair(publicKey, privateKey);
			}
			else
			{
				if (retryCount++ > MAX_RETRIES)
					throw MslInternalException("Too many retries while generating RSA key pair");
			}
		}
	}
}

shared_ptr<ByteArray> deriveWrappingKey(shared_ptr<ByteArray> encryptionKey, shared_ptr<ByteArray> hmacKey)
{
    // Derive Kpw key data from Kpe and Kph, according to the following algorithm:
    // Kpw = trunc_128(HMAC-SHA256(HMAC-SHA256(salt, cat(Kpe, Kph)), info))
    // where the first argument to HMAC-SHA256 is the key.
    // salt = 02 76 17 98 4f 62 27 53 9a 63 0b 89 7c 01 7d 69
    // info = 80 9f 82 a7 ad df 54 8d 3e a9 dd 06 7f f9 bb 91
    ensureOpenSslInit();
    try {

        const ByteArray salt(begin(saltAry), end(saltAry));
        const ByteArray info(begin(infoAry), end(infoAry));

        // cat Kpe and Kph
        ByteArray catK(*encryptionKey);
        catK.insert(catK.end(), hmacKey->begin(), hmacKey->end());

        // first HMAC(salt, catK)
        ByteArray sig1;
        signHmacSha256(salt, catK, sig1);

        // second HMAC(first HMAC, info)
        ByteArray sig2;
        signHmacSha256(sig1, info, sig2);

        // truncation and final output
         assert(sig2.size() > 128/8);
         shared_ptr<ByteArray> wrappingKey = make_shared<ByteArray>(sig2.begin(), sig2.begin() + 128/8);
         return wrappingKey;

    } catch (const MslCryptoException& e) {
        throw MslInternalException("Wrapping key derivation failed", e);
    }
}

shared_ptr<MasterToken> getMasterToken(shared_ptr<MslContext> ctx,
        int64_t sequenceNumber, int64_t serialNumber)
{
    shared_ptr<Date> renewalWindow = make_shared<Date>(Date::now()->getTime() + 10000);
    shared_ptr<Date> expiration = make_shared<Date>(Date::now()->getTime() + 20000);
    shared_ptr<EntityAuthenticationData> entityAuthData = ctx->getEntityAuthenticationData(MslContext::ReauthCode::ENTITYDATA_REAUTH);
    const string identity = entityAuthData->getIdentity();
    const SecretKey encryptionKey(MockPresharedAuthenticationFactory::KPE);
    const SecretKey hmacKey(MockPresharedAuthenticationFactory::KPH);
    return make_shared<MasterToken>(ctx, renewalWindow, expiration, sequenceNumber, serialNumber, make_shared<MslObject>(), identity, encryptionKey, hmacKey);
}

shared_ptr<MasterToken> getUntrustedMasterToken(shared_ptr<MslContext> ctx)
{
    shared_ptr<Date> renewalWindow = make_shared<Date>(Date::now()->getTime() + 10000);
    shared_ptr<Date> expiration = make_shared<Date>(Date::now()->getTime() + 20000);
    shared_ptr<EntityAuthenticationData> entityAuthData = ctx->getEntityAuthenticationData(MslContext::ReauthCode::INVALID);
    const string identity = entityAuthData->getIdentity();
    const SecretKey encryptionKey = MockPresharedAuthenticationFactory::KPE;
    const SecretKey hmacKey = MockPresharedAuthenticationFactory::KPH;
    shared_ptr<MasterToken> masterToken = make_shared<MasterToken>(ctx, renewalWindow, expiration, 1L, 1L, shared_ptr<io::MslObject>(), identity, encryptionKey, hmacKey);
    shared_ptr<MslEncoderFactory> encoder = ctx->getMslEncoderFactory();
    shared_ptr<MslObject> mo = toMslObject(encoder, masterToken);
    shared_ptr<ByteArray> signature = mo->getBytes("signature");
    ++(*signature)[1];
    mo->put("signature", signature);
    return make_shared<MasterToken>(ctx, mo);
}

shared_ptr<UserIdToken> getUserIdToken(shared_ptr<MslContext> ctx, shared_ptr<MasterToken> masterToken,
        int64_t serialNumber, shared_ptr<MslUser> user)
{
    shared_ptr<Date> renewalWindow = make_shared<Date>(Date::now()->getTime() + 10000);
    shared_ptr<Date> expiration = make_shared<Date>(Date::now()->getTime() + 20000);
    return make_shared<UserIdToken>(ctx, renewalWindow, expiration, masterToken, serialNumber, shared_ptr<MslObject>(), user);
}

shared_ptr<UserIdToken> getUntrustedUserIdToken(shared_ptr<MslContext> ctx,
        shared_ptr<MasterToken> masterToken, int64_t serialNumber,
        shared_ptr<MslUser> user)
{
    shared_ptr<Date> renewalWindow = make_shared<Date>(Date::now()->getTime() + 10000);
    shared_ptr<Date> expiration = make_shared<Date>(Date::now()->getTime() + 20000);
    shared_ptr<UserIdToken> userIdToken = make_shared<UserIdToken>(ctx, renewalWindow, expiration, masterToken, serialNumber, shared_ptr<MslObject>(), user);
    shared_ptr<MslEncoderFactory> encoder = ctx->getMslEncoderFactory();
    shared_ptr<MslObject> mo = toMslObject(encoder, userIdToken);
    shared_ptr<ByteArray> signature = mo->getBytes("signature");
    ++(*signature)[1];
    mo->put("signature", signature);
    return make_shared<UserIdToken>(ctx, mo, masterToken);
}

set<shared_ptr<ServiceToken>> getServiceTokens(shared_ptr<MslContext> ctx,
        shared_ptr<MasterToken> masterToken, shared_ptr<UserIdToken> userIdToken)
{
    shared_ptr<IRandom> random = ctx->getRandom();
    shared_ptr<ICryptoContext> cryptoContext = make_shared<NullCryptoContext>();
    set<shared_ptr<ServiceToken>> serviceTokens;
    int32_t numTokens = std::max(NUM_SERVICE_TOKENS, 3);
    for (int32_t i = 0; i < numTokens; ++i) {
        stringstream ss;
        ss << SERVICE_TOKEN_NAME << random->nextInt();
        const string name = ss.str();
        shared_ptr<ByteArray> data = make_shared<ByteArray>(32);
        random->nextBytes(*data);

        // Make sure one of each type of token is included.
        // Otherwise pick a random type.
        const shared_ptr<MasterToken> nullMasterToken;
        const shared_ptr<UserIdToken> nullUserIdToken;
        const int32_t type = (i < 3) ? i : random->nextInt(3);
        switch (type) {
            case 0:
                serviceTokens.insert(make_shared<ServiceToken>(ctx, name, data, nullMasterToken, nullUserIdToken, false, MslConstants::CompressionAlgorithm::NOCOMPRESSION, cryptoContext));
                break;
            case 1:
                serviceTokens.insert(make_shared<ServiceToken>(ctx, name, data, masterToken, nullUserIdToken, false, MslConstants::CompressionAlgorithm::NOCOMPRESSION, cryptoContext));
                break;
            case 2:
                serviceTokens.insert(make_shared<ServiceToken>(ctx, name, data, masterToken, userIdToken, false, MslConstants::CompressionAlgorithm::NOCOMPRESSION, cryptoContext));
                break;
            default:
                assert(false);
        }
    }
    return serviceTokens;
}

set<shared_ptr<ServiceToken>> getMasterBoundServiceTokens(shared_ptr<MslContext> ctx,
		shared_ptr<MasterToken> masterToken)
{
	shared_ptr<IRandom> random = ctx->getRandom();
	shared_ptr<ICryptoContext> cryptoContext = make_shared<NullCryptoContext>();
	set<shared_ptr<ServiceToken>> tokens;
	for (int count = random->nextInt(NUM_SERVICE_TOKENS); count >= 0; --count) {
        stringstream ss;
        ss << SERVICE_TOKEN_NAME << random->nextInt();
        const string name = ss.str();
		shared_ptr<ByteArray> data = make_shared<ByteArray>(8);
		random->nextBytes(*data);
		shared_ptr<ServiceToken> token = make_shared<ServiceToken>(ctx, name, data, masterToken, shared_ptr<UserIdToken>(), false, MslConstants::CompressionAlgorithm::NOCOMPRESSION, cryptoContext);
		tokens.insert(token);
	}
	return tokens;
}

set<shared_ptr<ServiceToken>> getUserBoundServiceTokens(shared_ptr<MslContext> ctx,
		shared_ptr<MasterToken> masterToken, shared_ptr<UserIdToken> userIdToken)
{
	shared_ptr<IRandom> random = ctx->getRandom();
	shared_ptr<ICryptoContext> cryptoContext = make_shared<NullCryptoContext>();
	set<shared_ptr<ServiceToken>> tokens;
	for (int count = random->nextInt(NUM_SERVICE_TOKENS); count >= 0; --count) {
        stringstream ss;
        ss << SERVICE_TOKEN_NAME << random->nextInt();
        const string name = ss.str();
		shared_ptr<ByteArray> data = make_shared<ByteArray>(8);
		random->nextBytes(*data);
		shared_ptr<ServiceToken> token = make_shared<ServiceToken>(ctx, name, data, masterToken, userIdToken, false, MslConstants::CompressionAlgorithm::NOCOMPRESSION, cryptoContext);
		tokens.insert(token);
	}
	return tokens;
}

}}}} // namespace netflix::msl::MslTestUtils
